### ISSUE TITLE
Semantic highlighting: `readonly` should depend on context

### ISSUE THREAD

**Issue by t-troebst** [clangd/clangd#1585]
I recently started experimenting with the semantic highlighting in clangd and specifically with distinguishing const and non-const variables. It seems that `readonly` only depends on the type in the declaration and not on the context, see the example below:

![image](https://user-images.githubusercontent.com/15610942/230810863-1073f834-cc63-474c-9698-b2fa9a94e582.png)

Here I have underlined every token that receives `readonly`.  Note that `s.x` in `test()` and `x` in `Test::test_const` are *not* marked `readonly` even though they cannot be modified. I think this feature would be a lot more useful if `readonly`  took this context into account.

**Comment by HighCommander4** [comment]
See also #1070 for another suggested modification to the semantics of the `readonly` modifier (mostly orthogonal I think, just linking them together for discoverability).

**Comment by t-troebst** [comment]
Yes, I think that issue and this one point to the fact, that the semantics are well defined right now. They should line up with constness for it to make sense. I think the most natural semantics for C++ for an identifier `x` are:

1. If `x` is a variable or binding, mark `x` readonly iff `x` is *not* a modifiable lvalue.
2. If `x`  is a data member in an expression of form `x`, `(...).x` or `(...)->x`, mark `x` readonly iff that expression is *not* a modifiable lvalue.
3. If `x`  is a member function, mark `x` readonly iff `x` is marked const.
4. If `x` is a type mark `x` read only iff `x` is const (not as useful, but hey why not).
5. In other cases (freestanding function, concept specialization, non-type template parameter, enum constant) always mark `x` readonly (or potentially never if these identifiers are clearly distinguished by the semantic highlighting already).

**Comment by daiyousei-qz** [comment]
I'm currently disabling the readonly modifier by theming since I find the current modifier assignment misleading. If we could move forward with this discussion, I'm happy to help implement the new semantic. I feel the only change needed is clang::clangd::isConst predicate in SemanticHighlighting.cpp

**Comment by t-troebst** [comment]
I agree. I actually just submitted this change for review https://reviews.llvm.org/D148284. This just calls the existing isConst but on the type of the expression instead of the type used in the declaration. I would say this is a very non-controversial change. However, I also agree with your proposed change in #1070. I think we should try to stick as close as possible to `const` since that has a well-defined meaning in C++.

One note, above I said "modifiable lvalue" since I wasn't sure if the type is formally `const` in that context or not - I actually realized that this is not quite the right thing. Since for example a struct that has a const member is not considered a modifiable lvalue (since you can't assign to it) but I think we should not mark that as readonly. Whether the expression is `const` is already the right thing.

Side note, I've been experimenting a lot with this recently. I actually found clangd's custom `usedAsMutableReference` and `usedAsMutablePointer` modifiers to be potentially even more useful than `readonly`. However, these also need some work. I have a patch that improves support there for operators and member functions. The main difference between them is whether you're more tolerant of false-positives or false-negatives when it comes to highlighting mutation.
